File: frontend/lang3s/src/lib/auth/permissions.ts
import { createAccessControl } from "better-auth/plugins/access";
import {
  adminAc,
  defaultStatements,
  userAc,
} from "better-auth/plugins/admin/access";

export const UserRoles = [
  "admin",
  "user",
  "analyst",
  "modeller",
  "dataLoader",
] as const;

export const Permissions = [
  "user:create",
  "user:list",
  "user:set-role",
  "user:ban",
  "user:impersonate",
  "user:delete",
  "user:set-password",
  "user:get",
  "user:update",
  "session:list",
  "session:revoke",
  "session:delete",
  "project:create",
  "project:share",
  "project:update",
  "project:delete",
  "project:view",
  "data:load",
  "data:update",
  "model:create",
  "model:delete",
  "model:export",
  "jobs:view",
  "jobs:delete",
  "jobs:create",
  "metadata:edit",
  "ontology:edit",
] as const;

export type UserRole = (typeof UserRoles)[number];
export type Permission = (typeof Permissions)[number];

const permissionsByAsset = {
  ...defaultStatements,
  project: ["create", "share", "update", "delete"],
  apiKey: ["create", "delete"],
  job: ["list", "get", "delete", "create"],
  metadata: ["edit"],
  ontology: ["edit"],
};

export const ac = createAccessControl(permissionsByAsset);

export const user = ac.newRole({
  ...userAc.statements,
});

export const admin = ac.newRole({
  project: ["create", "share", "update", "delete"],
  apiKey: ["create", "delete"],
  job: ["list", "get", "delete", "create"],
  metadata: ["edit"],
  ontology: ["edit"],
  ...adminAc.statements,
});

export const analyst = ac.newRole({
  project: ["create", "share", "update", "delete"],
  ...userAc.statements,
});

export const modeller = ac.newRole({
  project: ["create", "share", "update", "delete"],
  apiKey: ["create", "delete"],
  job: ["list", "get", "delete", "create"],
  metadata: ["edit"],
  ontology: ["edit"],
  ...userAc.statements,
});

export const dataLoader = ac.newRole({
  project: ["create", "share", "update", "delete"],
  apiKey: ["create", "delete"],
  job: ["list", "get", "delete", "create"],
  ...userAc.statements,
});

File: backend/packages/services/src/lang3s/services/permissions.py
from dataclasses import dataclass


@dataclass(frozen=True)
class PermissionAction:
    category: str
    value: str


@dataclass(frozen=True)
class PermissionCategory:
    pass


@dataclass(frozen=True)
class JobPermissions(PermissionCategory):
    list = PermissionAction("job", "list")
    create = PermissionAction("job", "create")
    delete = PermissionAction("job", "delete")
    get = PermissionAction("job", "get")


@dataclass(frozen=True)
class ApiKeyPermissions(PermissionCategory):
    create = PermissionAction("apiKey", "create")
    delete = PermissionAction("apiKey", "delete")


@dataclass(frozen=True)
class UserPermissions(PermissionCategory):
    create = PermissionAction("user", "create")
    list = PermissionAction("user", "list")
    set_role = PermissionAction("user", "set-role")
    ban = PermissionAction("user", "ban")
    impersonate = PermissionAction("user", "impersonate")
    impersonate_admin = PermissionAction("user", "impersonate-admin")
    delete = PermissionAction("user", "delete")
    set_password = PermissionAction("user", "set-password")
    set_email = PermissionAction("user", "set-email")
    get = PermissionAction("user", "get")
    update = PermissionAction("user", "update")


@dataclass(frozen=True)
class SessionPermissions(PermissionCategory):
    list = PermissionAction("session", "list")
    revoke = PermissionAction("session", "revoke")
    delete = PermissionAction("session", "delete")


@dataclass(frozen=True)
class ProjectPermissions(PermissionCategory):
    create = PermissionAction("project", "create")
    share = PermissionAction("project", "share")
    update = PermissionAction("project", "update")
    delete = PermissionAction("project", "delete")


@dataclass(frozen=True)
class MetadataPermissions(PermissionCategory):
    edit = PermissionAction("metadata", "edit")


@dataclass(frozen=True)
class OntologyPermissions(PermissionCategory):
    edit = PermissionAction("metadata", "edit")


class Permissions:
    job = JobPermissions()
    apiKey = ApiKeyPermissions()
    user = UserPermissions()
    session = SessionPermissions()
    project = ProjectPermissions()
    metadata = MetadataPermissions()
    ontology = OntologyPermissions()
